### ISSUE TITLE
[Performance] Autobot>Add slur test takes 20% more time than 4.0.2 build

### ISSUE THREAD

**Issue by DmitryArefiev** [musescore/MuseScore#17073]
### Issue type

Other type of issue

### Steps to reproduce

1. Run MU
2. Go to Menu panel>Diagnostic>Autobot>Show scripts
3. Run TC9: Big score (performance)
4. Check Add Slur time
5. Close MU without saving

Go through 1-5 steps three times and compare Add Slur results

4.0.2 release build Windows10

4.1 nightly master from 30 March

### MuseScore Version

4.1.0-230890304 nightly master

### Regression

Yes, this used to work in a previous version of Musescore 4.x

### Operating system

Win10/Mac13/Linux Ubuntu

**Comment by DmitryArefiev** [comment]
There is some performance decrease when adding slur (Autobot>**TC9: Big Score (performance)** test shows that). It could be related to engraving changes

**Comment by mike-spa** [comment]
Ok I've digged into the matter, and it is not related to engraving. The reason for the performance loss is that we enter the function `NotationInteraction::notifyAboutSelectionChangedIfNeed()` 3 times, while in 4.0.2 we only enter once. In turn, the reason we enter three times is that [here](https://github.com/musescore/MuseScore/pull/16431) we `setSelectionChanged(true)`. I'll leave it to @cbjeukendrup to investigate further, as I don't know the context of why that particular change was made (I can only say it seems totally logical to me). 

Also, I would point out that this is not a very good test for engraving performance. To really test engraving performance, I would suggest:

1. Using a real score, which has all kinds of different items (we can use our friendly Beethoven 9 score, for instance). The score used in this test is big, yes, but it's only notes without anything else: it's not going to put the engraving engine under stress, and it's visiting only a tiny portion of all the engraving functions.
2. Triggering a layout of the score in a different way, that doesn't involve selecting the entire score and adding items (which is also an unrealistic operation). Out of the total time it takes to perform this test, only a small part is spent for the layout of the score, and other operations such as managing and notifying the selection state become predominant. The easiest way I can think of to trigger a full score relayout is perhaps a style change, like changing the music font from Leland to Bravura. But maybe not even that: perhaps the purest test is just opening a file and monitoring the function `Score::doLayoutRange()`, which contains the whole engraving.

**Comment by DmitryArefiev** [comment]
> Also, I would point out that this is not a very good test for engraving performance. To really test engraving performance, I would suggest:
> 
> 1. Using a real score, which has all kinds of different items (we can use our friendly Beethoven 9 score, for instance). The score used in this test is big, yes, but it's only notes without anything else: it's not going to put the engraving engine under stress, and it's visiting only a tiny portion of all the engraving functions.
> 2. Triggering a layout of the score in a different way, that doesn't involve selecting the entire score and adding items (which is also an unrealistic operation). Out of the total time it takes to perform this test, only a small part is spent for the layout of the score, and other operations such as managing and notifying the selection state become predominant. The easiest way I can think of to trigger a full score relayout is perhaps a style change, like changing the music font from Leland to Bravura. But maybe not even that: perhaps the purest test is just opening a file and monitoring the function `Score::doLayoutRange()`, which contains the whole engraving.


@mike-spa Good idea! I'll think about it and create a task for a new automated tests. Thanks for your investigation!

**Comment by DmitryArefiev** [comment]
@bkunda I tried some real life cases with adding slurs, hairpins, articulations to several staves on 1-2 pages and didn't find a noticeable difference between **4.0.2** and **4.1 master**.

Let's bump it

**Comment by DmitryArefiev** [comment]
Tested on **4.2 beta1 build**, Win/Mac - **looks good** now. Roman says it's because of improvements on Properties side.

**4.2** Win:
5735
5837
5798
**4.1.1** Win:
7951
8124
7486

**4.2** Mac:
3550
3536
3609
**4.1.1** Win:
4575
5019
4823

Closing
